assembly/src/tokens/stream.rs

```diff
@@ -25,24 +26,40 @@ impl<'a> TokenStream<'a> {
             return Err(ParsingError::empty_source());
         }
         let mut tokens = Vec::new();
-        let mut doc_comments = BTreeMap::new();
+        let mut proc_comments = BTreeMap::new();
+        let mut module_comment = None;
 
-        let mut comment = Comment(None);
+        let mut comment_builder = CommentBuilder(None);
 
         for line in source.lines() {
             let line = line.trim();
             if line.starts_with(DOC_COMMENT_PREFIX) {
-                comment.append_line(line);
-            } else if line.is_empty() && !comment.is_empty() {
-                return Err(ParsingError::malformed_doc_comment(tokens.len()));
+                comment_builder.append_line(line);
+            } else if line.is_empty()
+                && !comment_builder.is_empty()
+                && tokens.is_empty()
+                && module_comment.is_none()
+            {
+                // If we haven't read any tokens yet, but already have built a comment, a new line
+                // must indicate the end of a module comment.
+                module_comment = comment_builder.take_comment();
             } else {
-                if !comment.is_empty() {
-                    doc_comments.insert(tokens.len(), comment.take_content());
-                }
                 let mut line_tokens = line
                     .split_whitespace()
                     .take_while(|&token| !token.starts_with(LINE_COMMENT_PREFIX))
                     .collect::<Vec<_>>();
+
+                if !comment_builder.is_empty() {
+                    // procedure comment should always be followed by a procedure token
+                    if !line_tokens.is_empty()
+                        && (line_tokens[0].split('.').collect::<Vec<&str>>()[0] == "export"
+                            || line_tokens[0].split('.').collect::<Vec<&str>>()[0] == "proc")
```

Comment:
Is this better than just doing `starts_with()`? Also, we instead of hardcoding "export" and "proc", we should use constants. We already have them under `Token::EXPORT` and `Token::PROC`.